src/CodeGenerator/src/Generator/ObjectGenerator.php

```diff
@@ -76,11 +89,50 @@ public function generate(StructureShape $shape): ClassName
         $this->addProperties($shape, $class, $namespace);
         $this->generateValidate($shape, $class, $namespace);
 
+        $serializer = $this->serializer->get($shape->getService());
+        if ($this->isShapeUsedInput($shape)) {
+            [$returnType, $requestBody, $args] = $serializer->generateRequestBuilder($shape) + [null, null, []];
+            $method = $class->addMethod('requestBody')->setReturnType($returnType)->setBody($requestBody)->setPublic()->setComment('@internal');
+            foreach ($args as $arg => $type) {
+                $method->addParameter($arg)->setType($type);
+            }
+        }
+
         $this->fileWriter->write($namespace);
 
         return $className;
     }
 
+    private function isShapeUsedInput(StructureShape $shape): bool
+    {
+        if (null === $this->usedShapedInput) {
+            $service = $shape->getService();
+            $walk = function (Shape $shape) use (&$walk) {
+                $this->usedShapedInput[$shape->getName()] = true;
+                if ($shape instanceof StructureShape) {
+                    foreach ($shape->getMembers() as $member) {
+                        $walk($member->getShape());
+                    }
+                } elseif ($shape instanceof ListShape) {
+                    $walk($shape->getMember()->getShape());
+                } elseif ($shape instanceof MapShape) {
+                    $walk($shape->getValue()->getShape());
+                }
+            };
```

Comment:
This will trap us in an infinite loop if structures contain eachother. I'll provide a test to show it.

Reply:
oh yes, I forgot the exit statement

```diff
            $walk = function (Shape $shape) use (&$walk) {
+               if (isset($this->usedShapedInput[$shape->getName()]) return;
                $this->usedShapedInput[$shape->getName()] = true;
                if ($shape instanceof StructureShape) {
                    foreach ($shape->getMembers() as $member) {
                        $walk($member->getShape());
                    }
                } elseif ($shape instanceof ListShape) {
                    $walk($shape->getMember()->getShape());
                } elseif ($shape instanceof MapShape) {
                    $walk($shape->getValue()->getShape());
                }
            };
```